### PR TITLE
perf(core): cache provider statuses for faster successive startup

### LINKED ISSUE #1824: garden exec is slow
### Current Behavior

When I run `garden exec php bash`, garden spends an awful lot of time "Getting status". Even after it's done there's a very long pause before dropping me into my shell. Pretty sure docker on its own was way quicker, unless kubernetes adds some kind of overhead?

### Expected behavior

Much quicker to get an interactive shell.

### Reproducible example

For my test I'm using a docker image built on `php:7.1-fpm`. 

### Workaround

*Maybe* there's a way to bypass garden to get into the container? I don't know what the exact command is though, that's why I'm using garden 😋 

### Suggested solution(s)

Make it fassttrrrr. Maybe spend less time probing the current state of things and just try executing the command? If that fails, *then* run some diagnosis?

### Your environment

`garden version` 0.11.13

`kubectl version` Client Version: version.Info{Major:"1", Minor:"16+", GitVersion:"v1.16.6-beta.0", GitCommit:"e7f962ba86f4ce7033828210ca3556393c377bcc", GitTreeState:"clean", BuildDate:"2020-01-15T08:26:26Z", GoVersion:"go1.13.5", Compiler:"gc", Platform:"windows/amd64"}
Server Version: version.Info{Major:"1", Minor:"16+", GitVersion:"v1.16.6-beta.0", GitCommit:"e7f962ba86f4ce7033828210ca3556393c377bcc", GitTreeState:"clean", BuildDate:"2020-01-15T08:18:29Z", GoVersion:"go1.13.5", Compiler:"gc", Platform:"linux/amd64"}

`docker version` 19.03.8

### PR DESCRIPTION
**Pull request by edvald**
**What this PR does / why we need it**:

We now cache provider statuses for a default of one hour. You can run
any command with `--force-refresh` to skip the caching, and can override
the cache duration with the `GARDEN_CACHE_TTL=<seconds>` environment
variable.

This substantially improves command execution times when run in
succession, which will be very useful for day-to-day usage, as well CI
performance.

**Which issue(s) this PR fixes**:

Closes #1824

**Special notes for your reviewer**:

It would be good to get some manual testing in on this one, try breaking it, look for edge cases etc.

### CODE REVIEW

File: garden-service/src/tasks/resolve-provider.ts [hunk 1]
```diff
@@ -197,13 +219,87 @@ export class ResolveProviderTask extends BaseTask {
...
+  private async getCachedStatus(config: ProviderConfig): Promise<EnvironmentStatus | null> {
+    const cachePath = this.getCachePath()
+
+    this.log.silly(`Checking provider status cache for ${this.plugin.name} at ${cachePath}`)
+
+    let cachedStatus: CachedStatus | null = null
+
+    if (!this.forceRefresh) {
+      try {
+        const cachedData = deserialize(await readFile(cachePath))
```
[the remark this answers is not in the document]
Comment (thsig) [reply garden-service/src/tasks/resolve-provider.ts:244]: I'm also wondering if we should get rid of the `cryo` library (which we use for serializing `ConfigMap`s for test & task results) and use these functions there instead.

File: garden-service/src/garden.ts [hunk 2]
```diff
@@ -634,6 +640,15 @@ export class Garden {
 
       providers = Object.values(taskResults).map((result) => result!.output)
 
+      let gotCachedResult = false
+
+      for (const provider of providers) {
```
Comment (thsig) [inline garden-service/src/garden.ts:645]: Could be replaced with:
```typescript
const gotCachedResult = !!providers.find((p) => p.status.cached)
```

File: garden-service/src/tasks/resolve-provider.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (edvald) [reply garden-service/src/tasks/resolve-provider.ts:244]: We should indeed drop `cryo` but that's for another PR. It didn't cross my mind to add unit tests for Node.js standard lib functions, what would you think is worth testing specifically?

File: garden-service/src/garden.ts (hunk 2 above)
Comment (edvald) [reply garden-service/src/garden.ts:645]: True. In a previous iteration there was more happening in the loop, so this got left in like this.

File: garden-service/src/tasks/resolve-provider.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (thsig) [reply garden-service/src/tasks/resolve-provider.ts:244]: Yeah, on second thought this would probably be overkill.

**Comment by edvald** [comment]
@thsig updated!

**Pull request merged**